### PR TITLE
Ruby Wrappers for NTheory

### PR DESCRIPTION
**Pull request by rajithv**

### CODE REVIEW

File: ext/symengine/ruby_ntheory.c [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+VALUE cntheory_quotient(VALUE self, VALUE operand1, VALUE operand2) {
+    return cntheory_twoarg(ntheory_quotient, operand1, operand2);
+}
+
+VALUE cntheory_fibonacci(VALUE self, VALUE operand1){
+    basic_struct *cresult;
+    VALUE result;
+
+    unsigned long cbasic_operand1;
+    cbasic_operand1 = NUM2ULONG(operand1);
```
Comment (rajithv) [inline ext/symengine/ruby_ntheory.c]: @abinashmeher999 am I using this correctly? Pls check.

File: ext/symengine/ruby_ntheory.c [hunk 2]
```diff
@@ -0,0 +1,105 @@
+#include "ruby_ntheory.h"
+
+VALUE cntheory_onearg(void (*cwfunc_ptr)(basic_struct*, const basic_struct*), VALUE operand1) {
```
Comment (isuruf) [inline ext/symengine/ruby_ntheory.c]: Isn't this the same as `cfunction_onearg`?

Comment (rajithv) [inline ext/symengine/ruby_ntheory.c]: @isuruf yes.. it is. 
Should I use that instead? 
Or should I make another source file with methods common to different parts?

What's the better way?

Comment (isuruf) [inline ext/symengine/ruby_ntheory.c]: Add it to `symengine_macros.h` and rename the file to `symengine_utils.h`

File: ext/symengine/ruby_ntheory.c (hunk 1 above)
Comment (abinashmeher999) [inline ext/symengine/ruby_ntheory.c]: Yes, that is correct. If you get unsure about return types, refer here: http://rxr.whitequark.org/mri/ident

File: ext/symengine/ruby_ntheory.c (hunk 2 above)
Comment (abinashmeher999) [inline ext/symengine/ruby_ntheory.c]: +1

File: spec/ntheory_spec.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      context 'NextPrime after 4' do
+        it 'returns 5' do
+          f = SymEngine::nextprime(4)
+          expect(f).to eql(5)
+        end
+      end
+    end
+
+    describe '#mod' do
+      context '5 mod 4' do
```
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: Also add tests for other permutations of a positive and a negative number and see if they behave the same way as ruby, because I have heard they [do not behave the same way in python and C](http://stackoverflow.com/questions/1907565/c-python-different-behaviour-of-the-modulo-operation).

File: ext/symengine/symengine_utils.c [hunk 4]
```diff
@@ -111,3 +111,40 @@ VALUE Klass_of_Basic(const basic_struct *basic_ptr) {
...
+
+    cresult = basic_new_heap();
+    cwfunc_ptr(cresult, cbasic_operand1);
+    result = Data_Wrap_Struct(Klass_of_Basic(cresult), NULL , cbasic_free_heap, cresult);
+    basic_free_stack(cbasic_operand1);
+
+    return result;
+}
+
+VALUE function_twoarg(void (*cwfunc_ptr)(basic_struct*, const basic_struct*, const basic_struct*), VALUE operand1, VALUE operand2) {
```
Comment (abinashmeher999) [inline ext/symengine/symengine_utils.c]: This line is too long. Break it if possible.

File: ext/symengine/symengine_utils.h [hunk 5]
```diff
@@ -2,12 +2,17 @@
...
 #include "symengine/cwrapper.h"
 
 //Returns the pointer wrapped inside the Ruby VALUE
 void sympify(VALUE operand2, basic_struct *cbasic_operand2);
 //Returns the Ruby class of the corresponding basic_struct pointer
 VALUE Klass_of_Basic(const basic_struct *basic_ptr);
+//Returns the result from the function pointed by cwfunc_ptr: for one argument functions
+VALUE function_onearg(void (*cwfunc_ptr)(basic_struct*, const basic_struct*), VALUE operand1);
+//Returns the result from the function pointed by cwfunc_ptr: for two argument functions
+VALUE function_twoarg(void (*cwfunc_ptr)(basic_struct*, const basic_struct*, const basic_struct*), VALUE operand1, VALUE operand2);
```
Comment (abinashmeher999) [inline ext/symengine/symengine_utils.h]: Same for this.

File: spec/ntheory_spec.rb (hunk 3 above)
Comment (rajithv) [inline spec/ntheory_spec.rb]: @abinashmeher999 The same happens in Ruby. So I should adapt the code to return Ruby-like answers. right?

**Comment by rajithv** [comment]
@abinashmeher999 @isuruf good to merge?

File: spec/ntheory_spec.rb (hunk 3 above)
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: IMHO symengine's behavior should be consistent across all languages. But again there shouldn't be any surprises for ruby users. If both are same then it's good.

File: ext/symengine/ruby_ntheory.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <ruby.h>
+#include <symengine/cwrapper.h>
+
+#include "symengine.h"
+#include "ruby_basic.h"
+#include "symengine_utils.h"
+
+VALUE cntheory_onearg(void (*cwfunc_ptr)(basic_struct*, const basic_struct*), VALUE operand1);
+VALUE cntheory_twoarg(void (*cwfunc_ptr)(basic_struct*, const basic_struct*, const basic_struct*), VALUE operand1, VALUE operand2);
```
Comment (abinashmeher999) [inline ext/symengine/ruby_ntheory.h]: Please break this one too.

**Comment by abinashmeher999** [comment]
Let's wait for the tests to pass.

File: spec/ntheory_spec.rb (hunk 3 above)
Comment (isuruf) [inline spec/ntheory_spec.rb]: I agree with the 2nd not the 1st point.
There are two versions, `mod` and `mod_f` and they are used in C++ and Python respectively. You want the `mod_f` version for Ruby

File: spec/ntheory_spec.rb [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        it 'returns 1' do
+          f = SymEngine::mod(5, 4)
```
Comment (isuruf) [inline spec/ntheory_spec.rb]: This should use ruby's `%` operator

File: ext/symengine/ruby_ntheory.c [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    return result;
+}
+
+VALUE cntheory_lucas(VALUE self, VALUE operand1){
+    basic_struct *cresult;
+    VALUE result;
+
+    unsigned long cbasic_operand1;
+    cbasic_operand1 = NUM2ULONG(operand1);
```
Comment (isuruf) [inline ext/symengine/ruby_ntheory.c]: What happens when `operand1` is a SymEngine integer?

File: ext/symengine/ruby_ntheory.h (hunk 6 above)
Comment (rajithv) [inline ext/symengine/ruby_ntheory.h]: It's actually not needed. Removed it.

File: ext/symengine/ruby_ntheory.c (hunk 8 above)
Comment (rajithv) [inline ext/symengine/ruby_ntheory.c]: now, it gives an error `TypeError: no implicit conversion of SymEngine::Integer into Integer`

Can I use `sympify` for this as well? In C++, the method signature is `SymEngine::lucas(unsigned long n)`. So the cwrapper also expects a unsinged long.

File: ext/symengine/ruby_ntheory.c (hunk 1 above)
Comment (isuruf) [inline ext/symengine/ruby_ntheory.c]: Then all you need to do is to add an `implicit conversion of SymEngine::Integer into Integer`

File: spec/ntheory_spec.rb (hunk 7 above)
Comment (rajithv) [inline spec/ntheory_spec.rb]: So I should overload % right? not introduce a new method as mod.

Comment (isuruf) [inline spec/ntheory_spec.rb]: Yes.

File: spec/ntheory_spec.rb (hunk 3 above)
Comment (rajithv) [inline spec/ntheory_spec.rb]: I modified the ruby wrapper to return the same as mod_f. 
But should I go ahead and wrap mod_f in SymEngine, and get that to the ruby wrapper? Or is it not necessary?

Also, should there be a way to access the c-style mod operation in Ruby?

Comment (isuruf) [inline spec/ntheory_spec.rb]: It'll be better to have the `mod_f` function in SymEngine. For now, that's okay. You can open an issue to keep track of it.

File: ext/symengine/ruby_ntheory.c (hunk 1 above)
Comment (rajithv) [inline ext/symengine/ruby_ntheory.c]: I figured our half way on how to do this:

so I added a new method to the 'Integer' class
'rb_define_method(c_integer, "to_int", cinteger_to_int, 0);'

From which I should return a VALUE pointer pointing to a Ruby `Bignum` or a `Fixnum`. This can be done by converting a C int by using `INT2NUM` macro.

The step I'm stuck at is on extracting a C int from the SymEngine Integer object. @abinashmeher999 can you point me to somewhere I can read more about the Ruby Wrapper's SymEngine Integer's structure?

Comment (certik) [inline ext/symengine/ruby_ntheory.c]: @rajithv, the Integer object uses GMP underneath. So you can extract a gmp int easily using the `as_mpz()` method. If you want an integer, you can call `as_int()`, but it will raise an exception if the integer doesn't fit into `signed long int`.

Comment (isuruf) [inline ext/symengine/ruby_ntheory.c]: You can convert a SymEngine::Integer to a Ruby string first and then call `to_i` method

Comment (abinashmeher999) [inline ext/symengine/ruby_ntheory.c]: +1. This can be done. So instead of using the Ruby C API for the above, you can define `#to_int` in ruby itself [here](https://github.com/symengine/symengine.rb/tree/master/lib/symengine). Make a new file `integer.rb` and then proceed.

File: spec/ntheory_spec.rb [hunk 9]
```diff
@@ -36,25 +36,25 @@
     describe '#mod' do
       context '5 mod 4' do
         it 'returns 1' do
-          f = SymEngine::mod(5, 4)
+          f = 5 % 4
```
Comment (isuruf) [inline spec/ntheory_spec.rb]: You are testing ruby's in built ints here

File: ext/symengine/symengine.c [hunk 10]
```diff
@@ -145,4 +146,14 @@ void Init_symengine() {
     rb_define_module_function(m_symengine, "zeta", cfunction_zeta, 1);
     rb_define_module_function(m_symengine, "gamma", cfunction_gamma, 1); 
 
+    //NTheory Functions as Module Level Functions
+    rb_define_module_function(m_symengine, "gcd", cntheory_gcd, 2);
+    rb_define_module_function(m_symengine, "lcm", cntheory_lcm, 2);
+    rb_define_module_function(m_symengine, "nextprime", cntheory_nextprime, 1);
+    rb_define_module_function(m_symengine, "%", cntheory_mod, 2);
```
Comment (abinashmeher999) [inline ext/symengine/symengine.c]: Defining this as a module level function won't work. You are not going to use it like `SymEngine::%(5, 4)` right?

File: spec/ntheory_spec.rb [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      end
+      context '5 mod -4' do
+        it 'returns -3' do
+          f = 5 % -4
+          expect(f).to eql(-3)
+        end
+      end
+      context '-5 mod 4' do
+        it 'returns 3' do
+          f = -5 % 4
```
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: None of these tests for `#mod` have `SymEngine` operands.

**Comment by rajithv** [comment]
@abinashmeher999 @isuruf pls review.

File: spec/ntheory_spec.rb [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+          f = SymEngine::quotient(5, 2)
+          expect(f).to eql(2)
+        end
+      end
+    end
+
+    describe '#fibonacci' do
+      context '5th Fibonacci Number' do
+        it 'returns 5' do
+          f = SymEngine::fibonacci(5)
```
Comment (isuruf) [inline spec/ntheory_spec.rb]: You should add another test which uses a SymEngine integer instead of Ruby's

File: spec/ntheory_spec.rb [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+          expect(f).to eql(5)
+        end
+      end
+    end
+
+    describe '#mod' do
+
+      i4 = SymEngine::Integer.new(4)
+      i5 = SymEngine::Integer.new(5)
+      i_4 = SymEngine::Integer.new(-4)
```
Comment (isuruf) [inline spec/ntheory_spec.rb]: Use `im4` and `im5`

File: spec/ntheory_spec.rb [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+          f = SymEngine::fibonacci(5)
+          expect(f).to eql(5)
+        end
+      end
+    end
+
+    describe '#lucas' do
+      context '1st Lucas Number' do
+        it 'returns 1' do
+          f = SymEngine::fibonacci(1)
```
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: Shouldn't this be `lucas`?

File: spec/ntheory_spec.rb (hunk 12 above)
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: Better to add it to all of them. Or at least one for one argument and one for two argument function.

**Comment by rajithv** [comment]
@isuruf @abinashmeher999 done

File: spec/ntheory_spec.rb [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+      context '5 mod -4' do
+        it 'returns -3' do
+          f = @i5 % @im4
+          expect(f).to eql(-3)
+        end
+      end
+      context '-5 mod 4' do
+        it 'returns 3' do
+          f = @im5 % @i4
+          expect(f).to eql(3)
```
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: So, `f` is a `SymEngine::Integer`, right? Will implicit conversion happen here while comparing?

File: spec/ntheory_spec.rb [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    end
+    
+    describe '#binomial' do
+      context 'binomial (n=5, k=1)' do
+        it 'returns 5' do
+          f = SymEngine::binomial(5, 1)
+          expect(f).to eql(5)
+        end
+      end
+      context 'binomial (n=5, k=1)' do
```
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: Make the wording of the context like `with SymEngine::Integer as parameters` and same for the test above this `with ruby Integer as parameters`. This is easier to read when tests are run in verbose mode. Currently it will print
```
#binomial
    binomial (n=5, k=1)
        returns 5
    binomial (n=5, k=1)
        returns 5
```
Accordingly modify `it` string. If the tests pass please make sure to make these changes in a new PR.

File: spec/ntheory_spec.rb (hunk 15 above)
Comment (abinashmeher999) [inline spec/ntheory_spec.rb]: So, `eql` checks [hash equality](http://stackoverflow.com/questions/7156955/whats-the-difference-between-equal-eql-and) and [type conversions are taken care in `eq`](https://www.relishapp.com/rspec/rspec-expectations/v/2-0/docs/matchers/equality-matchers).

**Comment by abinashmeher999** [comment]
@rajithv Thanks for the PR and your patience. +1 to merge. The minor changes can be incorporated in another PR.
